baseplate/clients/sqlalchemy.py

```diff
@@ -226,6 +226,7 @@ def on_before_execute(
 
         trace_name = f"{context_name}.execute"
         span = server_span.make_child(trace_name)
+        span.set_tag("protocol", "sqlalchemy")
```

Comment:
The more I think about it, the less I like using the protocol in this way, especially since "protocol" isn't even really descriptive here since "TCP" is just as valid as "SQL". 

Rather than having all metrics classes defined in `prometheus_metrics.py` I would like to see a new function on spans for setting the metrics class and then having it defined with where spans are being opened. In a sense, move the metrics class here, then use `span.set_metrics_cls(PrometheusSqlClientMetrics)`

That said, metrics class are also overly verbose.
1. I think that if you don't call `set_metrics_cls` that it should use a dynamically generated metrics class (as I had in my PR that was reverted out)
2. I like the functionality in metrics_tagged.py where certain tags are whitelisted to become labels on the metrics, which I think would reduce the verbosity here of needing to define a metrics class for every time we want metrics. This has the (perhaps positive) side effect of requiring the metrics classes be instantiated, which offers an initialization point where we can do things like setting the gauges and counters with the prefix rather than having to make new classes with class fields referring to the counters+gauges.

Reply:
I like something like this a lot more, and should be straight forward to do. I was away when that PR got introduced and reverted, can you describe what went wrong?

Reply:
🤔 Some thinking out loud alternatives:

1) (Current) Set a "protocol" tag on every span and have special cases in a generic Prometheus observer.

Weird dispatcher in the Prom observer, metric classes need to know about client/server implementation details like what tags are there and what they mean.

2) (This) Ditch the dispatching Prometheus observer and instead have Prom observer classes specific to a client or server (or subset of those). 

This acknowledges the tight coupling between client/server details and prometheus, and gets rid of the dispatcher. 

Would we want to keep the de facto interface we currently have on metric classes of `active_requests_metric`, `latency_seconds_metric` and so on? Are we sure it always makes sense, looking at the current things we have in Wavefront?

We'll probably still end up with lots of classes that are very similar except for some details like what tags to look at, or latency distribution. Instead of passing a metric class here we could pass an instance of a class, and then potentially change things like the distribution buckets from the client/server class without making a new metric class.

```
span.set_metrics_cls(PrometheusSqlClientMetrics(latency_buckets={...}))
```

3) Take it a step further and not use the observer pattern, and just build prom into the client/server classes.

Easy, less code, can target what we actually need. Introduces a specific dependency on prom being imported (even if it does nothing when prom is not enabled) which some people don't like? Spans outside baseplate would need to implement prom themselves (good?!).

4) Keep the dispatching Prometheus observer, remove the protocol tag and instead inspect the call stack to see where the span is from.

Keeps the span/observer abstraction clean, not needing to change every client and server. Still reliant on the prom classes knowing implementation details of the clients, probably a bit too magic.

Reply:
> metric classes need to know about client/server implementation details like what tags there are and what they mean
interesting I've always seen it the other way

The metric class dictates what you need to expose. The client is responsible to make sure to expose these to you. Not the other way around.